core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala

```diff
@@ -473,6 +473,326 @@ class DAGSchedulerSuite
+    val shuffleTwoRdd = new MyRDD(sc, 2, List(shuffleDepOne)).cache()
+    val shuffleDepTwo = new ShuffleDependency(shuffleTwoRdd, null)
+    val finalRdd = new MyRDD(sc, 1, List(shuffleDepTwo))
+    submit(finalRdd, Array(0))
+
+    // In the first two iterations, Stage 0 succeeds and stage 1 fails. In the next two iterations,
+    // stage 2 fails.
+    for (attempt <- 0 until Stage.MAX_STAGE_FAILURES) {
+      // Complete all the tasks for the current attempt of stage 0 successfully
+      val stage0Attempt = taskSets.last
+      checkStageId(0, attempt, stage0Attempt)
+      // Run stage 0
+      complete(stage0Attempt, makeCompletions(stage0Attempt, 2))
+
+      if (attempt < Stage.MAX_STAGE_FAILURES/2) {
+        // Now we should have a new taskSet, for a new attempt of stage 1.
+        // We will have one fetch failure for this task set
+        val stage1Attempt = taskSets.last
+        checkStageId(1, attempt, stage1Attempt)
+
+        val stage1Successes =
+          stage1Attempt.tasks.tail.map { _ => (Success, makeMapStatus("hostB", 1))}
+
+        // Run Stage 1, this time with a task failure
+        complete(stage1Attempt,
+          Seq((FetchFailed(makeBlockManagerId("hostA"),
+              shuffleDepOne.shuffleId, 0, 0, "ignored"), null)
+          ) ++ stage1Successes
+        )
+      } else {
+        // Run stage 1
+        val stage1Attempt = taskSets.last
+        checkStageId(1, attempt, stage1Attempt)
+        complete(stage1Attempt, makeCompletions(stage1Attempt, 1))
+
+        // Fail stage 2
+        val stage2Attempt = taskSets.last
+        checkStageId(2, attempt-Stage.MAX_STAGE_FAILURES/2, stage2Attempt)
+        complete(stage2Attempt, Seq(
+          (FetchFailed(makeBlockManagerId("hostA"),
+            shuffleDepTwo.shuffleId, 0, 0, "ignored"), null)))
+      }
+
+      // this will (potentially) trigger a resubmission of stage 0, since we've lost some of its
+      // map output, for the next iteration through the loop
+      scheduler.resubmitFailedStages()
+    }
+
+    // Complete all three stages succesfully
+    val stage0Attempt4 = taskSets.last
+    checkStageId(0, 4, stage0Attempt4)
+    complete(stage0Attempt4, makeCompletions(stage0Attempt4, 2))
+
+    val stage1Attempt4 = taskSets.last
+    checkStageId(1, 4, stage1Attempt4)
+    complete(stage1Attempt4, makeCompletions(stage1Attempt4, 1))
+
+    println(taskSets.mkString(","))
+
+    val stage2Attempt = taskSets.last
+    checkStageId(2, Stage.MAX_STAGE_FAILURES/2, stage2Attempt)
+    complete(stage2Attempt, Seq((Success, 42)))
+
+    // The first success is from the success we append in stage 1, the second is the one we add here
+    assert(results === Map(0 -> 42))
+  }
+
+  /**
+   * In this test we simulate a job failure where a stage may have many tasks, many of which fail.
+   * We want to show that many fetch failures inside a single stage do not trigger an abort on
+   * their own, but only when the stage fails enough times.
```

Comment:
can you reword slightly to make the attempts more clear : "... fetch failures inside a single stage attempt do not trigger an abort on their own, but only when there are enough failing stage attempts."